osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -70,6 +72,9 @@ public IEnumerable<BeatmapSetInfo> Beatmaps
 
         private readonly List<BeatmapGroup> groups = new List<BeatmapGroup>();
 
+        private Bindable<SelectionRandomType> randomType;
+        private readonly HashSet<BeatmapGroup> seenGroups = new HashSet<BeatmapGroup>();

```

Comment:
HashSet in order to better show that the order in which groups have been shown is not important and that there should be no duplicates. It is not strictly a necessity though, C# should be able to run O(M+N) set difference on lists too, right?